logstash-core/src/main/java/org/logstash/plugin/Input.java

```diff
@@ -0,0 +1,44 @@
+package org.logstash.plugin;
+
+import org.logstash.Event;
+
+import java.util.Collection;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+
+/**
+ * A Logstash input plugin.
+ * <p>
+ * Input plugins produce events intended given to Processors.
+ * <p>
+ * Inputs will generally run forever until there is some terminating condition such as a pipeline shutdown.
+ */
+public interface Input {
+    /**
+     * The method contract is as follows:
+     * <p>
+     * 1) When `Thread.interrupted()` is true, you MUST close all open resources and return.
+     * 2) When this method returns, it is assumed all work for this input is completed.
+     * 3) consumer.accept(...) may block
+     * <p>
+     * Acknowledging Data:
+     * <p>
+     * When `consumer.accept(...)` returns, the given Events have been successfully written into the consumer, and it is now
+     * safe to acknowledge these events to the upstream data source. The consumer is generally the Logstash internal queue,
+     * such as the persistent queue.
+     *
+     * @param consumer Send batches of events with consumer.accept(batch).
+     */
+    void run(Consumer<Collection<Event>> consumer);
+
+    /**
+     * Return this Input as a Callable for use with an ExecutorService.
```

Comment:
@jordansissel I think our inputs and workers are one of those rare cases where you rather want an explicit `Thread` per instance.
JRuby (and also in some spots we ourselves) is making heavy use of `Threadlocal` and using more `Threadlocal` logic for reusable buffers etc. would help performance greatly.
=> If we're planning on reusing threads across something like pipeline reloads then `Threadlocal` may be a problematic pattern => we don't want to reuse threads => `ExecutorService` doesn't really make things all that much simpler unless you want scheduling or thread reuse imo :)